nf_core/pipeline-template/{{cookiecutter.name_noslash}}/lib/Completion.groovy

```diff
@@ -0,0 +1,155 @@
+        def sendmail_html          = sendmail_template.toString()
+
+        // Send the HTML e-mail
+        Map colors = Headers.log_colours(params.monochrome_logs)
+        if (email_address) {
+            try {
+                if (params.plaintext_email) { throw GroovyException('Send plaintext e-mail, not HTML') }
+                // Try to send HTML e-mail using sendmail
+                [ 'sendmail', '-t' ].execute() << sendmail_html
+                log.info "-${colors.purple}[$workflow.manifest.name]${colors.green} Sent summary e-mail to $email_address (sendmail)-"
+            } catch (all) {
+                // Catch failures and try with plaintext
+                def mail_cmd = [ 'mail', '-s', subject, '--content-type=text/html', email_address ]
+                if ( mqc_report.size() <= max_multiqc_email_size.toBytes() ) {
+                    mail_cmd += [ '-A', mqc_report ]
+                }
+                mail_cmd.execute() << email_html
+                log.info "-${colors.purple}[$workflow.manifest.name]${colors.green} Sent summary e-mail to $email_address (mail)-"
+            }
+        }
+
+        // Write summary e-mail HTML to a file
+        def output_d = new File("${params.outdir}/pipeline_info/")
+        if (!output_d.exists()) {
+            output_d.mkdirs()
+        }
+        def output_hf = new File(output_d, "pipeline_report.html")
+        output_hf.withWriter { w -> w << email_html }
+        def output_tf = new File(output_d, "pipeline_report.txt")
+        output_tf.withWriter { w -> w << email_txt }
+    }
+
+    static void summary(workflow, params, log, fail_percent_mapped=[:], pass_percent_mapped=[:]) {
+        Map colors = Headers.log_colours(params.monochrome_logs)
+
+        if (pass_percent_mapped.size() > 0) {
+            def idx = 0
+            def samp_aln = ''
+            def total_aln_count = pass_percent_mapped.size() + fail_percent_mapped.size()
+            for (samp in pass_percent_mapped) {
+                samp_aln += "    ${samp.value}%: ${samp.key}\n"
+                idx += 1
+                if (idx > 5) {
+                    samp_aln += "    ..see pipeline reports for full list\n"
+                    break;
+                }
+            }
+            log.info "-${colors.purple}[$workflow.manifest.name]${colors.green} ${pass_percent_mapped.size()}/$total_aln_count samples passed STAR ${params.min_mapped_reads}% mapped threshold:\n${samp_aln}${colors.reset}-"
+        }
+        if (fail_percent_mapped.size() > 0) {
+            def samp_aln = ''
+            for (samp in fail_percent_mapped) {
+                samp_aln += "    ${samp.value}%: ${samp.key}\n"
+            }
+            log.info "-${colors.purple}[$workflow.manifest.name]${colors.red} ${fail_percent_mapped.size()} samples skipped since they failed STAR ${params.min_mapped_reads}% mapped threshold:\n${samp_aln}${colors.reset}-"
+        }
```

Comment:
Specific to `rnaseq` again. Would be great if we could have some kind of plugin system where pipelines can define their own functions to be injected in at various points like this? Then we can keep this file super generic and identical between pipelines.

Alternatively, just strip all of this and we are less strict about checking it in the linting.